File: web/domain/ocrAmountWeights.js
export const FEATURE_COUNT = 5;
export const DOLLAR_SIGN = 0;
export const STRONG_TOTAL_LABEL = 1;
export const TOTAL_LABEL = 2;
export const BOTTOM_HALF = 3;
export const ORDER_OR_POINTS_PENALTY = 4;

export const DEFAULTS = Object.freeze([50, 80, 60, 25, -100]);
export const STEP = 25;
export const CLAMP_MIN = 0;
export const CLAMP_MAX = 120;
export const PENALTY_MIN = -150;
export const PENALTY_MAX = 0;

export function fromDefaults() {
  return DEFAULTS.slice();
}

export function copyOrDefault(weights) {
  if (!weights || weights.length !== FEATURE_COUNT) return fromDefaults();
  return weights.slice();
}

export function encodeWeights(weights) {
  const w = copyOrDefault(weights);
  const buf = new ArrayBuffer(FEATURE_COUNT * 4);
  const view = new DataView(buf);
  for (let i = 0; i < FEATURE_COUNT; i++) view.setFloat32(i * 4, w[i], true);
  return new Uint8Array(buf);
}

export function decodeWeights(blob) {
  if (!blob || blob.length !== FEATURE_COUNT * 4) return fromDefaults();
  try {
    const view = new DataView(blob.buffer, blob.byteOffset, blob.byteLength);
    const out = [];
    for (let i = 0; i < FEATURE_COUNT; i++) out.push(view.getFloat32(i * 4, true));
    return out;
  } catch {
    return fromDefaults();
  }
}

export function clampWeights(weights) {
  const w = copyOrDefault(weights);
  for (let i = 0; i < ORDER_OR_POINTS_PENALTY; i++) {
    w[i] = Math.min(CLAMP_MAX, Math.max(CLAMP_MIN, w[i]));
  }
  w[ORDER_OR_POINTS_PENALTY] = Math.min(PENALTY_MAX, Math.max(PENALTY_MIN, w[ORDER_OR_POINTS_PENALTY]));
  return w;
}

File: app/src/main/java/com/example/envelopemoney/receipt/OcrAmountWeights.java
package com.example.envelopemoney.receipt;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

/**
 * Fixed-length OCR fallback weight vector stored as a little-endian float32 blob.
 *
 * <p>Index order: dollarSign, strongTotalLabel, totalLabel, bottomHalf, orderOrPointsPenalty.
 * Defaults match {@link ReceiptFieldParser} fallback scoring constants.
 */
public final class OcrAmountWeights {

    public static final int FEATURE_COUNT = 5;
    public static final int DOLLAR_SIGN = 0;
    public static final int STRONG_TOTAL_LABEL = 1;
    public static final int TOTAL_LABEL = 2;
    public static final int BOTTOM_HALF = 3;
    public static final int ORDER_OR_POINTS_PENALTY = 4;

    public static final float STEP = 25f;
    public static final float CLAMP_MIN = 0f;
    public static final float CLAMP_MAX = 120f;
    public static final float PENALTY_MIN = -150f;
    public static final float PENALTY_MAX = 0f;

    private static final float[] DEFAULTS = new float[]{50f, 80f, 60f, 25f, -100f};

    private OcrAmountWeights() {
    }

    public static float[] defaults() {
        return Arrays.copyOf(DEFAULTS, FEATURE_COUNT);
    }

    public static float[] copyOrDefault(float[] weights) {
        if (weights == null || weights.length != FEATURE_COUNT) {
            return defaults();
        }
        return Arrays.copyOf(weights, FEATURE_COUNT);
    }

    public static byte[] toLittleEndianBlob(float[] weights) {
        float[] w = copyOrDefault(weights);
        ByteBuffer buffer = ByteBuffer.allocate(FEATURE_COUNT * 4).order(ByteOrder.LITTLE_ENDIAN);
        for (int i = 0; i < FEATURE_COUNT; i++) {
            buffer.putFloat(w[i]);
        }
        return buffer.array();
    }

    /**
     * Decodes a 20-byte little-endian blob. Null, short, or unreadable data returns defaults.
     */
    public static float[] fromLittleEndianBlob(byte[] blob) {
        if (blob == null || blob.length != FEATURE_COUNT * 4) {
            return defaults();
        }
        try {
            ByteBuffer buffer = ByteBuffer.wrap(blob).order(ByteOrder.LITTLE_ENDIAN);
            float[] out = new float[FEATURE_COUNT];
            for (int i = 0; i < FEATURE_COUNT; i++) {
                out[i] = buffer.getFloat();
            }
            return out;
        } catch (Exception ignored) {
            return defaults();
        }
    }

    public static float[] clamp(float[] weights) {
        float[] w = copyOrDefault(weights);
        for (int i = 0; i < ORDER_OR_POINTS_PENALTY; i++) {
            w[i] = Math.min(CLAMP_MAX, Math.max(CLAMP_MIN, w[i]));
        }
        w[ORDER_OR_POINTS_PENALTY] = Math.min(PENALTY_MAX, Math.max(PENALTY_MIN, w[ORDER_OR_POINTS_PENALTY]));
        return w;
    }
}
